include/kllvm/binary/deserializer.h

```diff
@@ -210,68 +210,172 @@ class proof_trace_ringbuffer : public proof_trace_buffer {
+      return;
+    }
 
-      if (read_eof_) {
-        return false;
+    sem_wait(data_avail_);
+
+    if (shm_buffer_->eof()) {
+      // EOF has been written to the ringbuffer. Check if this is the last chunk
+      // of data to be read.
+      size_t remaining_data_sz = shm_buffer_->data_size();
+      if (remaining_data_sz < shm_ringbuffer::buffered_access_sz) {
+        // This is the last chunk of data to be read from the ringbuffer.
+        shm_buffer_->get(buffer_.data(), remaining_data_sz);
+        sem_post(space_avail_);
+
+        buffer_data_size_ = remaining_data_sz;
+        buffered_eof_ = true;
+
+        return;
       }
+    }
+
+    // Else, either EOF has not been written or EOF has been written but there
+    // are remaining full chunks to be read. In any case, we can read a full
+    // chunk.
+    shm_buffer_->get(buffer_.data());
+    sem_post(space_avail_);
+
+    buffer_data_size_ = shm_ringbuffer::buffered_access_sz;
+  }
+
+  bool read(uint8_t *ptr, size_t len = 1) {
+    // Check if we have read EOF already.
+    if (read_eof_) {
+      return false;
+    }
+
+    // Consume peeked data.
+    while (len > 0 && !peek_data_.empty()) {
+      *ptr = peek_data_.front();
+      peek_data_.pop_front();
+      ptr++;
+      len--;
+    }
 
-      while (int wait_status = sem_trywait(data_avail_)) {
-        assert(wait_status == -1 && errno == EAGAIN);
-        if (shm_buffer_->eof()) {
+    // Consume peeked EOF.
+    if (len > 0 && peek_eof_) {
+      read_eof_ = true;
+      return false;
+    }
+
+    // Peeked data has been fully consumed. If more data is requested, we need
+    // to read from buffer_ and/or shared memory.
+    while (len > 0) {
+      // If buffer_ is empty, try to fetch more data from the shared memory
+      // ringbuffer.
+      if (buffer_data_size_ == 0) {
+        // Check for and conusme buffered EOF.
+        if (buffered_eof_) {
           read_eof_ = true;
           return false;
         }
+
+        assert(buffer_data_start_ == 0);
+        read_from_shared_memory();
+      }
+
+      // Read available data from the buffer_.
+      assert(buffer_data_start_ < shm_ringbuffer::buffered_access_sz);
+      assert(
+          buffer_data_start_ + buffer_data_size_
+          <= shm_ringbuffer::buffered_access_sz);
+      size_t size_to_read_from_buffer = std::min(len, buffer_data_size_);
+      memcpy(
+          ptr, buffer_.data() + buffer_data_start_, size_to_read_from_buffer);
+      ptr += size_to_read_from_buffer;
+      len -= size_to_read_from_buffer;
+
+      buffer_data_start_ += size_to_read_from_buffer;
+      buffer_data_size_ -= size_to_read_from_buffer;
+      if (buffer_data_start_ == shm_ringbuffer::buffered_access_sz) {
+        assert(buffer_data_size_ == 0);
+        buffer_data_start_ = 0;
       }
-      shm_buffer_->get(&ptr[i]);
-      sem_post(space_avail_);
     }
 
+    assert(len == 0);
     return true;
   }
 
   bool peek(uint8_t *ptr, size_t len = 1) {
```

Comment:
Same question here

Reply:
This is just the wrapper, not the actual read to/from shared memory.

Reply:
As Dwight said, these are not the functions that do the actual shared memory accesses. They have default values of 1 because they are commonly called with this value and that makes the code more readable.